### ISSUE TITLE
Ansible is clunky and difficult to use in environments requiring an HTTP_PROXY for network access.

### ISSUE THREAD

**Issue by dpedu2** [ansible/ansible#22885]
##### ISSUE TYPE
 - Bug Report

But maybe also a Feature Idea?

##### COMPONENT NAME
Core

##### ANSIBLE VERSION
```
ansible 2.1.0.0
  config file = /home/ubuntu/xxx/ansible.cfg
  configured module search path = ['/home/ubuntu/xxx/library/']
```

##### CONFIGURATION
Not relevant

##### OS / ENVIRONMENT
Ubuntu 14.04

Network access must go through an http proxy, configured as typical in /etc/environment, via HTTP_PROXY and HTTPS_PROXY environment vars.

##### SUMMARY
Various modules that require network access (such as apt_key) are difficult to use in an environment such as the above. Ansible invokes shells in non-interactive mode which I understand does not load /etc/environment. 

A member on #ansible suggested http://docs.ansible.com/ansible/playbooks_environment.html. But this solution is strikingly manual for an "IT Automation" tool as it requires modifying every single play. Playbooks found online or through Galaxy must be manually modified as well. Additionally, there is NO workaround for ad-hoc commands - the plain `ansible` command does not provide a method for setting environment vars manually, or picking up any present on the remote system (short of falling back to the `shell` module for everything).

##### STEPS TO REPRODUCE
N/a

##### EXPECTED RESULTS
Ansible requires no special setup for machines behind http proxies.

##### ACTUAL RESULTS
Ansible's support for such an environment is nonexistent and the few available workarounds require manually specializing playbooks for such a case.

**Comment by abadger** [comment]
Set the environment variables in the shell's startup files.  (~/.profile, ~/.bashrc, ~/.zshrc, etc).

**Comment by dpedu2** [comment]
@abadger those startup files are not processed when the shells are started in non-interactive mode, which ansible uses for all its functions. That will not work.

**Comment by dpedu2** [comment]
For example:

```
ubuntu@import000:~$ echo 'export hello=1' | sudo tee -a /etc/profile
export hello=1
ubuntu@import000:~$ logout
Shared connection to import000 closed.
(bastion)ubuntu@ubuntu14:~/$ is import000
ubuntu@import000:~$ env | grep hello
hello=1
ubuntu@import000:~$ logout
Shared connection to import000 closed.
(bastion)ubuntu@ubuntu14:~/$ ansible import000.* -m shell -a 'env | grep hello'
import000.domain | FAILED | rc=1 >>
```

**Comment by bcoca** [comment]
@dpedu2 that depends on your system, there are other files that are always sourced (/etc/environment, /etc/conf.d/, /etc/profile, etc)

**Comment by dpedu2** [comment]
No, that's incorrect @bcoca. I suggest reading the man pages for sh or doing a simple experiment yourself:

```
$ ansible import001.* -m shell -b -a 'echo "hello_ansible=1234" | tee -a /etc/environment'
import001.daveazure9 | SUCCESS | rc=0 >>
hello_ansible=1234
$ ssh import001 env | grep hello
hello_ansible=1234
$ ssh import001 sudo reboot
$ ssh import001 env | grep hello
hello_ansible=1234
$ ansible import001.* -m shell -b -a 'env | grep hello'
import001.daveazure9 | FAILED | rc=1 >>
```

Please reopen my issue. The workarounds provided above are factually incorrect.

**Comment by bcoca** [comment]
I'll reiterate, @dpedu2 that depends on your system. Not all source the same files, i was giving you a list of  examples as this varies across OSs and even distros of the same OS.

**Comment by dpedu2** [comment]
@bcoca what platforms have you found this to be working on?

**Comment by larsks** [comment]
The request for giving the ad-hoc `ansible` command a way to set environment variables (maybe a `--environment VARNAME=VARVALUE` option or something) does not seem like an unreasonable request, since without that many modules will not operate correctly.

With `ansible-playbook`, the fact that the `enviroment:` key can be set on plays, blocks, and when using roles makes life manageable (since it doesn't need to be set per task).  This should work fine when using third-party roles from galaxy,  as in:

```- hosts: all
  environment:
    http_proxy: http://proxy.example.org:3128
  roles:
    - galaxy/role1
    - galaxy/role2
```

**Comment by dpedu2** [comment]
Using `enviroment:` is a workaround, yes. Though it is quite clunky and inefficient to need to paste that code over many playbooks; we can come up with a better solution than that. And what if your playbooks need to support hosts that do and do not use proxies? Sure, you could put the proxies in host's vars and pass that in plays again with the `environment:` statement. But this gets uglier and uglier for a simple problem.

**Comment by larsks** [comment]
@bcoca said:

> I'll reiterate, @dpedu2 that depends on your system. 

I don't think that's true.  The list of files that are sourced by, e.g., `bash` or whatever are typically *not* distribution specific.  A distribution may instrument something like `/etc/profile` to load in additional files, but I haven't seen one that would patch the shell.

A non-interactive shell will in general not source any dotfiles.  This true at least for `bash` and the traditional Bourne `sh`, so I think @abadger's suggestion was misinformed.

@dpedu2 said:

> what if your playbooks need to support hosts that do and do not use proxies?

Can you propose a model that you would consider an acceptable solution to the problem?

**Comment by dpedu2** [comment]
I think the most obvious answer would be to either create shells in a way where their startup files are used or to read the startup files manually. But this may not be realistic. I don't see a way to make `sh` read these files without running it with `-i` for interactive mode which drastically alters behavior and won't work.

Looking at similar tools, having support for an http proxy within configuration files of the automation solution itself seems to be a common built-in feature:

- https://docs.puppet.com/puppet/latest/configuration.html#httpproxyhost
- https://docs.saltstack.com/en/latest/ref/configuration/examples.html#example-minion-configuration-file
- https://docs.chef.io/proxies.html

This may be the way to go, but it still feels dirty to me. A matter of opinion - I don't really like the idea of a configuration file directive created specifically to set a single env var.

But I think this reveals the right direction - having the ability to set environmental vars for hosts at the inventory level. Since you're probably configuring the proxy setup of your target hosts using ansible plays too, I think it's fair to assume where ansible is being executed from has knowledge of these values.

How would this look? Ansible provides the [ability to set template/host variables from the inventory](http://docs.ansible.com/ansible/intro_inventory.html#group-variables). So I propose similar for environment vars - an inventory could look like:

```
[southeast:children]
atlanta
raleigh

[southeast:vars]
some_server=foo.southeast.example.com
halon_system_timeout=30
self_destruct_countdown=60
escape_pods=2

[southeast:envvars]
http_proxy=http://1.2.3.4:4567/
myproduct_debug_mode=1
foo=bar
```

**Comment by bcoca** [comment]
Environment is not only set by the files sourced by shell, its also set by your login, as well as any configured by ssh/sudo/su  as well as any that you inherit from your init system. These ARE mostly dependent on system configuration, distro/OS defaults, etc. The files sourced by your shell can also be configured, the ~/.bashrc is normally set only for interactive shells, but that also can be changed, I'm not going to go into detail for every possible system and or combination, you should check docs on your target systems to find out (if you dont want to read them, `ansible -m shell -a 'env' target ` should show you what you can grep for in /etc).

As for a better way to set environment, we have been looking at different options (config file and cli  included), the other factor is 'when set in 2 different places do we replace or merge?' ... better yet, find a way to let user decide. We have not been able to come up with something satisfactory yet. If you have a good idea, please let us know http://github.com/ansible/proposals.

**Comment by abadger** [comment]
A non-interactive shell will in general not source any dotfiles. This true
at least for bash and the traditional Bourne sh, so I think @abadger
<https://github.com/abadger>'s suggestion was misinformed.


Tested before posting with both bash and zsh (can't remember if I used a
fedora or rhel vm) so this definitely was an informed suggestion.  The
mechanism by which those systems make this happen I did not look into, just
that it works when you add the environment settings to the correct files.

**Comment by dpedu2** [comment]
Can you post how you tested it? Your statements conflict with the test commands I posted above.

**Comment by abadger** [comment]
Just tested on a variety of boxes and so far the following has worked on all of them.  My controller is Fedora 25 but I'm not sure that that matters (can't think of anything off the top of my head... The env var I'm testing is not set on the controller).

```
$ cat ~/.zshenv
ONE=zshenv
export ONE

$ head -2 ~/.bashrc
ONE=bashrc
export ONE
```

kill all active ssh muxs so that you don't end up with a previous ssh connection (otherwise you'll end up with the previous login session and therefore environment).
```
$ ansible -i 'fedora25-test,'  all -m setup>setup -e ansible_python_interpreter=/usr/bin/python3  
$ grep ONE setup                                                 (16:53:44)
            "ONE": "bashrc",
# chsh on the fedora25 box to zsh, kill all ssh muxes on the controller:
$ ansible -i 'fedora25-test,'  all -m setup>setup -e ansible_python_interpreter=/usr/bin/python3
$ grep ONE setup                                                 (16:55:39)
            "ONE": "zshenv",
```

On my centos7 and my ubuntu16 vms, ~/.zshenv and ~/.bashrc  also work in the same way.

Note that I hit upon .zshenv and .bashrc by adding the variable, ONE, to all of the startup files I thought might be sourced on startup and then seeing which one showed up in my ansible output.  So there could be more than just these files that do the trick.

**Comment by abadger** [comment]
Since this ticket was reported against Ubuntu 14.04 I just tested with my Ubuntu14 vm as well:
```
$ ansible -i 'ubuntu14-test,'  all -m setup>setup -e ansible_python_interpreter=/usr/bin/python3
$ grep ONE setup                                                 (17:12:38)
            "ONE": "zshenv",
$ # ssh to ubuntu14-test, chsh to /bin/bash, logout, kill all ssh muxes on the controller
$ ansible -i 'ubuntu14-test,'  all -m setup>setup -e ansible_python_interpreter=/usr/bin/python3
$ grep ONE setup                                                 (17:15:22)
            "ONE": "bashrc",
```

zsh's documentation suggests that .zshenv is used whenever zsh is invoked ( http://zsh.sourceforge.net/Intro/intro_3.html ) so I'm not especially surprised by that.  bash's behaviour is a bit more confusing... When I read the documentation the first time I would have assumed .profile or perhaps none of the files would be sourced.  But after testing the behaviour and reading the page again ( https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html ) my assumption would be that it falls under "Invoked by remote shell daemon" which is documented to use ~/.bashrc

Of course, it could still be using ~/.bashrc due to a distribution customization.  But it works on the three distros that I have immediate access to so if it a customization, it's not an obscure one.

**Comment by dpedu2** [comment]
What happens if you use `-b`, become?

**Comment by abadger** [comment]
Only tested on ubuntu14 with sudo, bash, and the root user.  There seem to be multiple ways to have become work depending on precisely the behaviour you wish.  You do have to modify the sudo_flags configuration parameter to make any of the ones I tried work (but I did not test this as exhaustively as the tests last week.   There may be other ways to do this.... In particular, zsh's documentation says that it utilized certain startup files whenever any shell is started so it may work by modifying those files without any further tweaks)

In the examples below, I've set the environment variable, ONE, in the remote_user's .bashrc.  I've set the environment variable, TWO, in the root account's .bashrc.

If you want to (and your security needs are amenable to) use the environment of the remote user to set http_proxy for the sudo'd user, add -E to ansible's sudo_flags:
```
$ grep sudo_flags /etc/ansible/ansible.cfg
sudo_flags = -H -S -n -E
$ ansible -i 'ubuntu14-test,'  all -m setup>setup -e ansible_python_interpreter=/usr/bin/python3 --sudo -K
SUDO password:
$ grep ONE setup
            "ONE": "bashrc", 
$ grep TWO setup
$
```

If you want to (and your security needs are amenable to) use the environment of the root user to set http_proxy for the sudo'd user, add -i to ansible's sudo_flags:
```
$ grep sudo_flags /etc/ansible/ansible.cfg
sudo_flags = -H -S -n -i
$ ansible -i 'ubuntu14-test,'  all -m setup>setup -e ansible_python_interpreter=/usr/bin/python3 --sudo -K
SUDO password:
$ grep ONE setup
$ grep TWO setup
            "TWO": "bashrc", 
```
